pkg/filewatcher/filewatcher.go

```diff
@@ -0,0 +1,253 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filewatcher
+
+import (
+	"bufio"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/hashicorp/go-multierror"
+)
+
+type FileWatcher interface {
+	Add(path string) error
+	Remove(path string) error
+	Close() error
+	Events(path string) chan fsnotify.Event
+	Errors(path string) chan error
+}
+
+type fsNotifyWatcher struct {
+	mu sync.RWMutex
+	// The watcher maintain a map of workers,
+	// keyed by watched dir (parent dir of watched files).
+	workers map[string]*worker
+}
+
+type worker struct {
+	// watcher is a fsnotify watcher that watches the parent
+	// dir of watchedFiles.
+	watcher *fsnotify.Watcher
+	// The worker maintain a map of event channel (included in fileTracker),
+	// keyed by watched file path.
+	// The worker watches parent path of given path,
+	// and filter out events of given path, then redirect
+	// to the result channel.
+	// Note that for symlink files, the content in received events
+	// do not have to be related to the file itself.
+	watchedFiles map[string]*fileTracker
+}
+
+type fileTracker struct {
+	events chan fsnotify.Event
+	// md5 sum to indicate if a file has been updated.
+	md5Sum string
+}
+
+// NewWatcher return with a FileWatcher instance that implemented with fsnotify.
+func NewWatcher() FileWatcher {
+	w := &fsNotifyWatcher{
+		workers: map[string]*worker{},
+	}
+
+	return w
+}
+
+// Add is implementation of Add interface of FileWatcher.
+func (w *fsNotifyWatcher) Add(path string) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	cleanedPath, parentPath := formatPath(path)
+	if worker, workerExist := w.workers[parentPath]; workerExist {
+		if _, pathExist := worker.watchedFiles[cleanedPath]; pathExist {
+			return fmt.Errorf("path %s already added", path)
+		}
+
+		// And the path into worker maps if not exist.
+		md5Sum, err := getMd5Sum(cleanedPath)
+		if err != nil {
+			return fmt.Errorf("failed to get md5 sum for %s: %v", path, err)
+		}
+		worker.watchedFiles[cleanedPath] = &fileTracker{
```

Comment:
Will this entry be orphaned if the subsequent errors in the method return an error?

Reply:
I think We'll return with nil error after this.
But I did forget to close newly created watcher on error, for places like L104